router/src/infer/tool_grammar.rs

```diff
@@ -20,44 +20,47 @@ impl ToolGrammar {
 
     pub fn apply(
         tools: Vec<Tool>,
-        tool_choice: ToolChoice,
+        tool_choice: ChatCompletionToolChoiceOption,
     ) -> Result<(Vec<Tool>, Option<JsonSchemaTool>), InferError> {
         // if no tools are provided, we return None
         if tools.is_empty() {
             return Ok((tools, None));
         }
 
-        let tool_choice = tool_choice.0.unwrap_or(ToolType::OneOf);
-
         let mut tools = tools.clone();
 
-        // add the no_tool function to the tools
-        let no_tool = Tool {
-            r#type: "function".to_string(),
-            function: FunctionDefinition {
-                name: "no_tool".to_string(),
-                description: Some("Open ened response with no specific tool selected".to_string()),
-                arguments: json!({
-                    "type": "object",
-                    "properties": {
-                        "content": {
-                            "type": "string",
-                            "description": "The response content",
-                        }
-                    },
-                    "required": ["content"]
-                }),
-            },
-        };
-        tools.push(no_tool);
+        // add the no_tool function to the tools as long as we are not required to use a specific tool
+        if tool_choice != ChatCompletionToolChoiceOption::Required {
+            let no_tool = Tool {
```

Comment:
Isn't this only ever used in the `auto` branch ? Wouldn't putting it in the match branch make more sense ?

Reply:
yep good point. Just moved into the match branch and removed the `let mut tools` and only mutate tools if the Auto option is chosen. thanks